src/JsonApiDotNetCore.MongoDb/MongoEntityRepository.cs

```diff
@@ -0,0 +1,182 @@
+        {
+            var resource = _resourceFactory.CreateInstance<TResource>();
+            resource.Id = id;
+
+            return Task.FromResult(resource);
+        }
+
+        public virtual Task CreateAsync(TResource resourceFromRequest, TResource resourceForDatabase,
+            CancellationToken cancellationToken)
+        {
+            if (resourceFromRequest == null) throw new ArgumentNullException(nameof(resourceFromRequest));
+            if (resourceForDatabase == null) throw new ArgumentNullException(nameof(resourceForDatabase));
+            
+            foreach (var attribute in _targetedFields.Attributes)
+            {
+                attribute.SetValue(resourceForDatabase, attribute.GetValue(resourceFromRequest));
+            }
+
+            return Collection.InsertOneAsync(resourceForDatabase, new InsertOneOptions(), cancellationToken);
+        }
+
+        public virtual async Task<TResource> GetForUpdateAsync(QueryLayer queryLayer, CancellationToken cancellationToken)
+        {
+            var resources = await GetAsync(queryLayer, cancellationToken);
+            return resources.FirstOrDefault();
+        }
+
+        public virtual async Task UpdateAsync(TResource requestResource, TResource databaseResource, CancellationToken cancellationToken)
+        {
+            foreach (var attr in _targetedFields.Attributes)
+                attr.SetValue(databaseResource, attr.GetValue(requestResource));
+
+            await Collection.ReplaceOneAsync(
+                Builders<TResource>.Filter.Eq(e => e.Id, databaseResource.Id),
+                databaseResource,
+                new ReplaceOptions(),
+                cancellationToken);
+        }
+
+        public virtual async Task DeleteAsync(TId id, CancellationToken cancellationToken)
+        {
+            var result = await Collection.DeleteOneAsync(
+                Builders<TResource>.Filter.Eq(e => e.Id, id),
+                new DeleteOptions(),
+                cancellationToken);
+
+            if (!result.IsAcknowledged || result.DeletedCount == 0)
+            {
+                throw new DataStoreUpdateException(new Exception());
+            }
+        }
+
+        public virtual Task SetRelationshipAsync(TResource primaryResource, object secondaryResourceIds, CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual Task AddToToManyRelationshipAsync(TId primaryId, ISet<IIdentifiable> secondaryResourceIds, CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual Task RemoveFromToManyRelationshipAsync(TResource primaryResource, ISet<IIdentifiable> secondaryResourceIds,
+            CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
+        }
+
+        protected virtual IMongoQueryable<TResource> ApplyQueryLayer(QueryLayer layer)
```

Comment:
Parameter null check is missing here, as well as in other public methods.
Also please move this method below CountAsync(), where it is being used for the first time. This keeps it more in sync with the main repo too.